CatalogReaderExample/NuGet.Protocol.Catalog/CatalogProcessor.cs

```diff
@@ -0,0 +1,207 @@
+                        pageItems.Count);
+                    break;
+                }
+            }
+        }
+
+        private async Task<bool> ProcessPageAsync(DateTimeOffset minCommitTimestamp, CatalogPageItem pageItem)
+        {
+            var page = await _client.GetPageAsync(pageItem.Url);
+
+            var leafItems = page.GetLeavesInBounds(
+                minCommitTimestamp,
+                _settings.MaxCommitTimestamp,
+                _settings.ExcludeRedundantLeaves);
+            _logger.LogInformation(
+                "On page {page}, {leaves} out of {totalLeaves} were in the time bounds.",
+                pageItem.Url,
+                leafItems.Count,
+                page.Items.Count);
+
+            DateTimeOffset? newCursor = null;
+            var success = true;
+            for (var i = 0; i < leafItems.Count; i++)
+            {
+                var leafItem = leafItems[i];
+
+                if (newCursor.HasValue && newCursor.Value != leafItem.CommitTimestamp)
+                {
+                    await _cursor.SetAsync(newCursor.Value);
+                }
+
+                newCursor = leafItem.CommitTimestamp;
+
+                success = await ProcessLeafAsync(leafItem);
+                if (!success)
+                {
+                    _logger.LogWarning(
+                        "{unprocessedLeaves} out of {leaves} leaves were left incomplete due to a processing failure.",
+                        leafItems.Count - i,
+                        leafItems.Count);
+                    break;
+                }
+            }
+
+            if (newCursor.HasValue && success)
+            {
+                await _cursor.SetAsync(newCursor.Value);
+            }
+
+            return success;
+        }
+
+        private async Task<bool> ProcessLeafAsync(CatalogLeafItem leafItem)
+        {
+            bool success;
+            try
+            {
+                switch (leafItem.Type)
+                {
+                    case CatalogLeafType.PackageDelete:
+                        var packageDelete = await _client.GetPackageDeleteLeafAsync(leafItem.Url);
+                        success = await _leafProcessor.ProcessPackageDeleteAsync(packageDelete);
+                        break;
+                    case CatalogLeafType.PackageDetails:
+                        var packageDetails = await _client.GetPackageDetailsLeafAsync(leafItem.Url);
+                        success = await _leafProcessor.ProcessPackageDetailsAsync(packageDetails);
+                        break;
+                    default:
+                        throw new NotSupportedException($"The catalog leaf type '{leafItem.Type}' is not supported.");
+                }
+            }
+            catch (Exception exception)
+            {
+                _logger.LogError(
+                    0,
+                    exception,
+                    "An exception was thrown while processing leaf {leaf}.",
```

Comment:
`{leafUrl}` may be clearer than just `{leaf}`.

Reply:
Fixed.